### PR TITLE
Get the file location based on path and conn_id

### PR DESCRIPTION
**Pull request by sunank200**
# Description
## What is the current behavior?

Currently `create_file_location` is picking module path based on `filetype.value` even tough the `conn_id` is passed. https://github.com/astronomer/astro-sdk/blob/main/python-sdk/src/astro/files/locations/__init__.py#L23 get_location_type doesn't take care of fetching the filelocation based on `conn_type` as per `conn_id` at all. It just picks up from the path. 

closes: #1471 

## What is the new behavior?

- Add `validate_conn()` in BaseFilelocation class to match file path and connection type

## Does this introduce a breaking change?
No

### Checklist
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

**Review by pankajastro — COMMENTED** [review]
do we need a test for this change?

File: python-sdk/src/astro/files/locations/base.py [hunk 1]
```diff
@@ -109,12 +119,24 @@ def check_non_existing_local_file_path(path: str) -> bool:
...
         :param path: Path to a file in the filesystem/Object stores
+        :param conn_id: Airflow connection id
         """
+        connection_type = ""
+        if conn_id is not None:
+            connection_type = BaseHook.get_connection(conn_id).conn_type
+
         file_scheme = urlparse(path).scheme
+
+        file_scheme_from_conn_id = CUSTOM_CONN_TYPE_TO_FILE_SCHEME.get(connection_type)
```
Comment (utkarsharma2) [inline python-sdk/src/astro/files/locations/base.py:134]: @sunank200 I think we should unit test case for this method. WDYT?

**Comment by utkarsharma2** [comment]
Also, I think, we can add this quirk to docs as well.

File: python-sdk/src/astro/files/locations/__init__.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def create_file_location(path: str, conn_id: Optional[str] = None) -> BaseFileLo
     :param path: Path to a file in the filesystem/Object stores
     :param conn_id: Airflow connection ID
     """
-    filetype: FileLocation = BaseFileLocation.get_location_type(path)
+    filetype: FileLocation = BaseFileLocation.get_location_type(path, conn_id)
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/__init__.py:22]: I don't think we should enforce getting location type from conn_id. Filepath is fine enough.

How we authorize to that filetype is an implementation detail

File: python-sdk/src/astro/files/locations/base.py [hunk 3]
```diff
@@ -72,14 +81,14 @@ def openlineage_dataset_name(self) -> str:
         raise NotImplementedError
 
     @staticmethod
-    def is_valid_path(path: str) -> bool:
+    def is_valid_path(path: str, conn_id: str | None = None) -> bool:
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/base.py:84]: I don't like this

File: python-sdk/src/astro/files/locations/__init__.py (hunk 2 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/__init__.py:22]: We are not using `conn_id` to determine the file location type. When the user explicitly passes `conn_id`  the intent here is that the file path and connection passed are aligned.

File: python-sdk/src/astro/files/locations/base.py (hunk 3 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/base.py]: Removed this as this is optional. https://github.com/astronomer/astro-sdk/pull/1478/commits/8f3e88f213d3fa87c82ba21251b258848fa0dcc4

File: python-sdk/src/astro/files/locations/__init__.py (hunk 2 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/__init__.py]: Changed the approach as discussed.

File: python-sdk/src/astro/files/locations/base.py [hunk 4]
```diff
@@ -109,12 +119,24 @@ def check_non_existing_local_file_path(path: str) -> bool:
...
+        if conn_id is not None:
+            connection_type = BaseHook.get_connection(conn_id).conn_type
+
         file_scheme = urlparse(path).scheme
+
+        file_scheme_from_conn_id = CUSTOM_CONN_TYPE_TO_FILE_SCHEME.get(connection_type)
+        if file_scheme_from_conn_id is not None and file_scheme_from_conn_id is not file_scheme:
+            raise ValueError(
+                f"Unsupported scheme '{file_scheme}' from path"
+                f" '{path}' in connection type '{connection_type}'"
```
[the remark this answers is not in the document]
Comment (sunank200) [reply python-sdk/src/astro/files/locations/base.py]: Have changed the error message

File: python-sdk/src/astro/files/locations/base.py (hunk 1 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/base.py]: added: https://github.com/astronomer/astro-sdk/pull/1478/commits/5a0ed3ff0b89713f7d39867754857b31fd01e423

**Comment by sunank200** [comment]
> do we need a test for this change?

@pankajastro Added the test: https://github.com/astronomer/astro-sdk/pull/1478/commits/5a0ed3ff0b89713f7d39867754857b31fd01e423 and changed the test respectively

File: python-sdk/docs/development/CONTRIBUTING.md [hunk 5]
```diff
@@ -79,5 +79,6 @@ Before you submit a pull request (PR), check that it meets these guidelines:
 
 -   Run tests locally before opening PR.
 
--   Adhere to guidelines for commit messages described in this [article](http://chris.beams.io/posts/git-commit/).
+-   Adhere to guidelines for commit messages described in this
+    [article](https://gist.github.com/tonibardina/9290fbc7d605b4f86919426e614fe692).
```
[the remark this answers is not in the document]
Comment (sunank200) [reply python-sdk/docs/development/CONTRIBUTING.md:83]: The previous URL is throwing 404.

File: python-sdk/tests/files/locations/test_location_base.py [hunk 6]
```diff
@@ -108,25 +109,20 @@ class SomethingElseLocation:  # skipcq: PY-D0002
 @pytest.mark.parametrize(
     "loc_1,loc_2,equality",
     [
-        (LocalLocation("/tmp/file_a.csv"), LocalLocation("/tmp/file_a.csv"), True),
+        (GCSLocation("gs://tmp/file_a.csv"), GCSLocation("gs://tmp/file_a.csv"), True),
         (
-            LocalLocation("/tmp/file_a.csv", conn_id="test"),
-            LocalLocation("/tmp/file_a.csv", conn_id="test"),
+            GCSLocation("gs://tmp/file_a.csv", conn_id="google_cloud_default"),
```
Comment (utkarsharma2) [inline python-sdk/tests/files/locations/test_location_base.py:114]: @sunank200 Why did we modify this test case?

File: python-sdk/src/astro/files/locations/base.py [hunk 7]
```diff
@@ -7,9 +7,19 @@
 from urllib.parse import urlparse
 
 import smart_open
+from airflow.hooks.base import BaseHook
 
 from astro.constants import FileLocation
 
+CUSTOM_CONN_TYPE_TO_FILE_SCHEME = {
+    "bigquery": {"gs"},
+    "google_cloud_platform": {"gs", "gdrive", "https"},
```
Comment (utkarsharma2) [inline python-sdk/src/astro/files/locations/base.py:16]: @sunank200 I think we don't support `https` paths as GCS path, right?

Comment (sunank200) [reply python-sdk/src/astro/files/locations/base.py:16]: So, this is the scenario with `gdrive`, it uses `https` paths but uses `google_cloud_platform` as connection_type

File: python-sdk/tests/files/locations/test_location_base.py (hunk 6 above)
Comment (sunank200) [reply python-sdk/tests/files/locations/test_location_base.py:114]: The `conn_id` is validated as part of BaseFileLocation call

File: python-sdk/src/astro/files/locations/base.py [hunk 8]
```diff
@@ -7,9 +7,19 @@
...
+    "s3": {"s3"},
+    "aws": {"s3"},
+    "wasb": {"wasb"},
+    "gcpbigquery": {"gs"},
+}
+
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/base.py:22]: This should instead go in the classes themselves.

```python
class GCSLocation:
     supported_conn_type = {GCSHook.conn_type}

```

File: python-sdk/docs/development/CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaxil) [reply python-sdk/docs/development/CONTRIBUTING.md:83]:
```suggestion
    [article](https://cbea.ms/git-commit/).
```

[the remark this answers is not in the document]
Comment (sunank200) [reply python-sdk/docs/development/CONTRIBUTING.md]: https://github.com/astronomer/astro-sdk/pull/1478/commits/513264dd45ef9e4d309f2401e9ede4b982cb8a07

File: python-sdk/src/astro/files/locations/base.py (hunk 7 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/base.py]: changed the logic here: https://github.com/astronomer/astro-sdk/pull/1478/commits/181366c944d65117e7ab0d8b06b72ffc9c23a270

File: python-sdk/src/astro/files/locations/base.py (hunk 8 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/base.py]: Changed the logic: https://github.com/astronomer/astro-sdk/pull/1478/commits/181366c944d65117e7ab0d8b06b72ffc9c23a270

File: python-sdk/src/astro/files/locations/http.py [hunk 9]
```diff
@@ -11,12 +11,18 @@ class HTTPLocation(BaseFileLocation):
...
+    # Google Drive uses HTTP location
+    supported_conn_type = {"google_cloud_platform"}
 
     @property
     def paths(self) -> list[str]:
         """Resolve patterns in path"""
         return [self.path]
 
+    def validate_conn(self):  # skipcq: PTC-W0049
+        """Override as conn_id is not required at all for http location."""
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/http.py:23]:
```suggestion
        """Override as conn_id is not always required http location."""
```

File: python-sdk/src/astro/files/locations/http.py [hunk 10]
```diff
@@ -11,12 +11,18 @@ class HTTPLocation(BaseFileLocation):
     """Handler http location operations"""
 
     location_type = FileLocation.HTTP
+    # Google Drive uses HTTP location
+    supported_conn_type = {"google_cloud_platform"}
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/http.py:15]: Do we even need to add this here if we are not validating the connection type for HTTP?

File: python-sdk/tests/files/locations/test_location_base.py [hunk 11]
```diff
@@ -142,3 +138,9 @@ def test_location_eq(loc_1, loc_2, equality):
 def test_location_hash():
     """Test that hashing works"""
     assert isinstance(hash(LocalLocation("/tmp/file_a.csv")), int)
+
+
+def test_invalid_conn_id_with_file_path():
+    """Raise a value when the connection types doesn't match the path"""
+    with pytest.raises(ValueError):
+        GCSLocation("gs://tmp/file_a.csv", conn_id="aws_default")
```
Comment (kaxil) [inline python-sdk/tests/files/locations/test_location_base.py:146]: fyi: you can pass a `match` param to `pytest.raises` to check for the error message:

Example:

```python
import pytest


def myfunc():
    raise ValueError("Exception 123 raised")


def test_match():
    with pytest.raises(ValueError, match=r".* 123 .*"):
        myfunc()
```

Comment (kaxil) [reply python-sdk/tests/files/locations/test_location_base.py:146]: From https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

File: python-sdk/tests/files/operators/test_files.py [hunk 12]
```diff
@@ -6,7 +6,8 @@
 from astro.files.operators.files import ListFileOperator
 
 
-def test_get_file_list_execute_local():
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
```
Comment (kaxil) [inline python-sdk/tests/files/operators/test_files.py:9]: Why do we need to mock `validate_conn`?

Comment (kaxil) [reply python-sdk/tests/files/operators/test_files.py:9]: Can we instead not pass the connection id on L15

File: python-sdk/tests/files/operators/test_files.py [hunk 13]
```diff
@@ -17,7 +18,8 @@ def test_get_file_list_execute_local():
 
 
 @patch("astro.files.locations.google.gcs.GCSLocation.hook")
-def test_get_file_list_execute_gcs(hook):
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
```
Comment (kaxil) [inline python-sdk/tests/files/operators/test_files.py:21]: Same comment as above

File: python-sdk/tests/files/operators/test_files.py [hunk 14]
```diff
@@ -30,7 +32,8 @@ def test_get_file_list_execute_gcs(hook):
 
 
 @patch("astro.files.locations.amazon.s3.S3Location.hook")
-def test_get_file_list_s3(hook):
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
```
Comment (kaxil) [inline python-sdk/tests/files/operators/test_files.py:35]: Same as above

Comment (sunank200) [reply python-sdk/tests/files/operators/test_files.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/50bd36d6d59208d166c01b6c00eec416cebc6e15)

File: python-sdk/tests/files/operators/test_files.py (hunk 13 above)
Comment (sunank200) [reply python-sdk/tests/files/operators/test_files.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/50bd36d6d59208d166c01b6c00eec416cebc6e15)

File: python-sdk/tests/files/operators/test_files.py (hunk 12 above)
Comment (sunank200) [reply python-sdk/tests/files/operators/test_files.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/50bd36d6d59208d166c01b6c00eec416cebc6e15)

File: python-sdk/tests/files/locations/test_location_base.py (hunk 11 above)
Comment (sunank200) [reply python-sdk/tests/files/locations/test_location_base.py]: Thanks. [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/50bd36d6d59208d166c01b6c00eec416cebc6e15)

File: python-sdk/src/astro/files/locations/http.py (hunk 10 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/http.py]: Removed it it was part of some local change. [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/50bd36d6d59208d166c01b6c00eec416cebc6e15)

File: python-sdk/src/astro/files/locations/http.py (hunk 9 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/http.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/50bd36d6d59208d166c01b6c00eec416cebc6e15)

**Comment by kaxil** [comment]
test is failing:

```
FAILED tests/files/operators/test_files.py::test_get_file_list_execute_gcs - airflow.exceptions.AirflowNotFoundException: The conn_id `bigquery` isn't defined

```

File: python-sdk/src/astro/files/locations/http.py [hunk 15]
```diff
@@ -17,6 +17,10 @@ def paths(self) -> list[str]:
         """Resolve patterns in path"""
         return [self.path]
 
+    def validate_conn(self):  # skipcq: PTC-W0049
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/http.py:20]: `# skipcq: PTC-W0049` shouldn't be needed if you have docstring

File: python-sdk/src/astro/files/locations/local.py [hunk 16]
```diff
@@ -25,6 +25,10 @@ def paths(self) -> list[str]:
             paths = glob.glob(url.path)
         return paths
 
+    def validate_conn(self):  # skipcq: PTC-W0049
+        """Override as conn_id is not always required for local location."""
+        return
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/local.py:30]: same as https://github.com/astronomer/astro-sdk/pull/1478/files#r1058450963

File: python-sdk/tests/files/type/test_base.py [hunk 17]
```diff
@@ -1,13 +1,17 @@
+from unittest.mock import patch
+
 from astro.constants import FileType
 from astro.files import File
 
 
-def test_str():
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
```
Comment (kaxil) [inline python-sdk/tests/files/type/test_base.py:7]: shouldn't be needed

File: python-sdk/tests/files/type/test_base.py [hunk 18]
```diff
@@ -1,13 +1,17 @@
...
+def test_str(validate_conn):
     file = File(path="astro", conn_id="local", filetype=FileType.CSV)
     assert file.__str__() == "astro"
 
 
-def test_repr():
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
```
Comment (kaxil) [inline python-sdk/tests/files/type/test_base.py:13]: shouldn't be needed

File: python-sdk/tests/files/type/test_base.py [hunk 19]
```diff
@@ -16,12 +20,14 @@ def test_repr():
...
-def test_eq():
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
+def test_eq(validate_conn):
     file1 = File(path="astro", conn_id="local", filetype=FileType.CSV)
     file2 = File(path="astro", conn_id="local", filetype=FileType.CSV)
     assert file1.__eq__(file2) is True
 
 
-def test_hash():
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
```
Comment (kaxil) [inline python-sdk/tests/files/type/test_base.py:30]: shouldn't be needed

File: python-sdk/tests/files/type/test_base.py [hunk 20]
```diff
@@ -16,12 +20,14 @@ def test_repr():
     )
 
 
-def test_eq():
+@patch("astro.files.locations.base.BaseFileLocation.validate_conn", return_value=None)
```
Comment (kaxil) [inline python-sdk/tests/files/type/test_base.py:23]: shouldn't be needed

File: python-sdk/src/astro/files/locations/http.py (hunk 15 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/http.py:20]: Screenshot 2022-12-28 at 5 03 56 PM
Actually, it was complaining even when I had doc-string

Comment (kaxil) [reply python-sdk/src/astro/files/locations/http.py:20]: try and remove the `return`, as it is redundant

File: python-sdk/src/astro/files/locations/http.py [hunk 21]
```diff
@@ -17,6 +17,10 @@ def paths(self) -> list[str]:
...
+        """Override as conn_id is not always required for http location."""
+        return
```
Comment (kaxil) [inline python-sdk/src/astro/files/locations/http.py:22]:
```suggestion
```

File: python-sdk/src/astro/files/locations/http.py (hunk 15 above)
Comment (kaxil) [inline python-sdk/src/astro/files/locations/http.py:20]:
```suggestion
    def validate_conn(self):
```

Comment (kaxil) [reply python-sdk/src/astro/files/locations/http.py]: lets see -- trying in https://github.com/astronomer/astro-sdk/pull/1478/commits/775f94713f280883cd169887064ebd60141e05dc

Comment (kaxil) [reply python-sdk/src/astro/files/locations/http.py]: yup deepsource passed for it :)

File: python-sdk/src/astro/files/locations/local.py (hunk 16 above)
Comment (kaxil) [inline python-sdk/src/astro/files/locations/local.py:30]:
```suggestion
```

File: python-sdk/tests/files/type/test_base.py (hunk 17 above)
Comment (sunank200) [reply python-sdk/tests/files/type/test_base.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/b89f84be043188126f642e50de139bf330adb72b)

File: python-sdk/tests/files/type/test_base.py (hunk 19 above)
Comment (sunank200) [reply python-sdk/tests/files/type/test_base.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/b89f84be043188126f642e50de139bf330adb72b)

File: python-sdk/tests/files/type/test_base.py (hunk 18 above)
Comment (sunank200) [reply python-sdk/tests/files/type/test_base.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/b89f84be043188126f642e50de139bf330adb72b)

File: python-sdk/tests/files/type/test_base.py (hunk 20 above)
Comment (sunank200) [reply python-sdk/tests/files/type/test_base.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/b89f84be043188126f642e50de139bf330adb72b)

File: python-sdk/src/astro/files/locations/local.py (hunk 16 above)
Comment (sunank200) [reply python-sdk/src/astro/files/locations/local.py]: [Done](https://github.com/astronomer/astro-sdk/pull/1478/commits/c7795fecefc2c88138d9a615c9f1b7a4d7316576)

**Review by kaxil — APPROVED** [review]

**Pull request merged**